snapshots/devmapper/pool_device.go

```diff
@@ -0,0 +1,323 @@
+type PoolDevice struct {
+	poolName string
+	metadata *PoolMetadata
+}
+
+// NewPoolDevice creates new thin-pool from existing data and metadata volumes.
+// If pool 'poolName' already exists, it'll be reloaded with new parameters.
+func NewPoolDevice(ctx context.Context, config *Config) (*PoolDevice, error) {
+	log.G(ctx).Infof("initializing pool device %q", config.PoolName)
+
+	version, err := dmsetup.Version()
+	if err != nil {
+		log.G(ctx).Errorf("dmsetup not available")
+		return nil, err
+	}
+
+	log.G(ctx).Infof("using dmsetup:\n%s", version)
+
+	dbpath := filepath.Join(config.RootPath, config.PoolName+".db")
+	poolMetaStore, err := NewPoolMetadata(dbpath)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := openPool(ctx, config); err != nil {
+		return nil, err
+	}
+
+	return &PoolDevice{
+		poolName: config.PoolName,
+		metadata: poolMetaStore,
+	}, nil
+}
+
+func openPool(ctx context.Context, config *Config) error {
+	if err := config.Validate(); err != nil {
+		return err
+	}
+
+	var (
+		poolPath   = dmsetup.GetFullDevicePath(config.PoolName)
+		poolExists = false
+	)
+
+	if _, err := os.Stat(poolPath); err != nil && !os.IsNotExist(err) {
+		return errors.Wrapf(err, "failed to stat for %q", poolPath)
+	} else if err == nil {
+		poolExists = true
+	}
+
+	// Create new pool if not exists
+	if !poolExists {
+		log.G(ctx).Debug("creating new pool device")
+		if err := dmsetup.CreatePool(config.PoolName, config.DataDevice, config.MetadataDevice, config.DataBlockSizeSectors); err != nil {
+			return errors.Wrapf(err, "failed to create thin-pool with name %q", config.PoolName)
+		}
+
+		return nil
+	}
+
+	// Pool exists, check if it needs to be reloaded
+	if config.DataDevice != "" && config.MetadataDevice != "" {
+		log.G(ctx).Debugf("reloading existing pool %q", poolPath)
+		if err := dmsetup.ReloadPool(config.PoolName, config.DataDevice, config.MetadataDevice, config.DataBlockSizeSectors); err != nil {
+			return errors.Wrapf(err, "failed to reload pool %q", config.PoolName)
+		}
+
+		return nil
+	}
+
+	// If data and meta devices are not provided, use existing pool. Query info to make sure it's OK.
+	if _, err := dmsetup.Info(poolPath); err != nil {
+		return errors.Wrapf(err, "failed to query info for existing pool %q", poolPath)
+	}
+
+	return nil
+}
+
+// transition invokes 'updateStateFn' callback to perform devmapper operation and reflects device state changes/errors in meta store.
+// 'tryingState' will be set before invoking callback. If callback succeeded 'successState' will be set, otherwise
+// error details will be recorded in meta store.
+func (p *PoolDevice) transition(ctx context.Context, deviceName string, tryingState DeviceState, successState DeviceState, updateStateFn func() error) error {
+	// Set device to trying state
+	uerr := p.metadata.UpdateDevice(ctx, deviceName, func(deviceInfo *DeviceInfo) error {
```

Comment:
If fail, the deviceID should not be reused, right? if so, could we clean it after reboot the containerd or somewhere?

Reply:
That's an interesting question :) Device ids are persistent across reboots, and there is something wrong with device id, would it be safe to reuse it after restart?

Reply:
The deviceID will be generated by `getNextDeviceID` method. And if the deviceID A is marked by `deviceFree`, the next call to createSnapshot can use the A as deviceID. However, the `transition` action might fail, for example, containerd reboot or the `updateStateFn` fails. That is odd case:). Since there are several transactions in `CreateThinDevice` method, if one fails, the data still is put into the metadata, for example, the deviceID turns into `Taken`. There are no rollback action in the `CreateThinDevice`. The next createSnapshot call will use new deviceID so that the id will be leak. :p . I think we should add rollback action for the devmapper's metadata.  

The `s.withTransaction` would not help because it is for storage metadata, not for the devmapper.

https://github.com/containerd/containerd/blob/b5f7aca209f8b51e55bc1ddce26d4d9b9584e1ed/snapshots/devmapper/snapshotter.go#L184-L198

Does it make senses?

Reply:
Fixed